### PR TITLE
Added appserviceplan and webapp commands

### PR DESCRIPTION
**Pull request by LukaszStem**
The content to be added to Changelog is as follows:
Added arm commands:
*   webapp config show
*   webapp config update
*   webapp publishprofile get
*   appserviceplan create
*   appserviceplan list
*   appserviceplan show
*   appserviceplan delete

Updated websites client api version 10.0 -> 11.0

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@LukaszStem__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (lukaszs). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

File: lib/commands/arm/appserviceplan/appserviceplan._js [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+          location: location,
+          properties: {
+            sku: sku,
+            numberOfWorkers: numberOfWorkers,
+            workerSize: workerSize
+          }
+        }
+      };
+      
+      log.info('---' + JSON.stringify(parameters));
```
Comment (amarzavery) [inline lib/commands/arm/appserviceplan/appserviceplan._js]: Why do you need this?

File: lib/commands/arm/appserviceplan/appserviceplan._js [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+  var appserviceplan = cli.category('appserviceplan')
+    .description($('Commands to manage your Azure appserviceplans'));
+
+  appserviceplan.command('create [resource-group] [name] [location] [sku] [numberOfWorkers] [workerSize]')
+    .description($('Create an appserviceplan'))
+    .usage('[options] <resource-group> <name> <location> <sku> <number-of-workers> <worker-size>')
+    .option('-g --resource-group <resource-group>', $('Name of the resource group'))
+    .option('-n --name <name>', $('the name of the appserviceplan to create'))
+    .option('-l --location <location>', $('the geographic region to create the appserviceplan'))
+    .option('-k --sku <sku>', $('the sku of the appserviceplan eg: Basic.'))
```
Comment (amarzavery) [inline lib/commands/arm/appserviceplan/appserviceplan._js]: can you provide the set of valid values in the help. Is Basic the only valid value?

File: lib/commands/arm/appserviceplan/appserviceplan._js [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+    .option('-w --workers <numberOfWorkers>', $('the number of workers in the appserviceplan'))
+    .option('-z --workersize <workerSize>', $('the size of workers in the appserviceplan'))
```
Comment (amarzavery) [inline lib/commands/arm/appserviceplan/appserviceplan._js]: How would the customer know what to provide? Big, small, xxl, medium? Please document the sizes in the help or provide a link to an MSDN doc that provides the valid values.

File: lib/commands/arm/appserviceplan/appserviceplan._js [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+            numberOfWorkers: numberOfWorkers,
+            workerSize: workerSize
+          }
+        }
+      };
+      
+      log.info('---' + JSON.stringify(parameters));
+      var result;
+      try {
+        result = client.webHostingPlans.createOrUpdate(resourceGroup, parameters, _);
```
Comment (amarzavery) [inline lib/commands/arm/appserviceplan/appserviceplan._js]: Why are you not using the latest version (0.11.0) of `azure-arm-website`. It does not contain webHostingPlans at all.

File: lib/commands/arm/appserviceplan/appserviceplan._js [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+        progress.end();
+      }
+      log.info(util.format($('appserviceplan %s has been deleted'), name));
+    });
+
+  appserviceplan.command('list [resource-group]')
+    .description($('List your appserviceplans'))
+    .usage('[options] <resource-group>')
+    .option('-g --resource-group <resource-group>', $('the name of the resource group'))
+	  .option('-s --subscription <subscription>', $('the subscription identifier'))
```
Comment (amarzavery) [inline lib/commands/arm/appserviceplan/appserviceplan._js]: indentation is incorrect

File: lib/commands/arm/appserviceplan/appserviceplan._js [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+      cli.interaction.formatOutput(result, function (data) {
+        if (!data) {
+          log.info($('No appserviceplan information available'));
+        } else {
+          log.data($('App Service Plan Name  :'), data.webHostingPlan.name);
+          log.data($('Location           :'), data.webHostingPlan.location);
+          log.data($('SKU                :'), data.webHostingPlan.properties.sku);
+          log.data($('Number of workers  :'), data.webHostingPlan.properties.numberOfWorkers);
+
+		  var workerSizeLiteral;
```
Comment (amarzavery) [inline lib/commands/arm/appserviceplan/appserviceplan._js]: indentation should be 2 spaces.

File: lib/commands/arm/appserviceplan/appserviceplanUtils._js [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+'use strict';
+
+var utils = require('../../../util/utils');
+
+exports.createWebappManagementClient = function (subscription) {
```
Comment (amarzavery) [inline lib/commands/arm/appserviceplan/appserviceplanUtils._js]: This entire file is not required.

File: lib/commands/arm/webapp/webapp.config._js [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+        .description($('Commands to manage your Azure webapps'));
+
+    var config = webapp.category('config')
+        .description($('Commands to manage your Azure webapps configurations'));
+
+    var publish = webapp.category('publishprofile')
+        .description($('Command to get your Azure webapps publishing profile'));
+
+
+    config.command('get [resource-group] [name]')
```
Comment (amarzavery) [inline lib/commands/arm/webapp/webapp.config._js]: command name should be `show`

File: lib/commands/arm/webapp/webapp._js [hunk 9]
```diff
@@ -290,4 +290,9 @@ exports.init = function (cli) {
         }
       });
     });
+
```
Comment (amarzavery) [inline lib/commands/arm/webapp/webapp._js]: Remove the empty lines

File: lib/commands/arm/appserviceplan/appserviceplanUtils._js (hunk 7 above)
Comment (amarzavery) [inline lib/commands/arm/appserviceplan/appserviceplanUtils._js]: Please take a look at the lib/util/utils.js file. It has methods for creating the client. Please add a new method over there if you dont see one for webapp.

File: lib/commands/arm/webapp/webapp.config._js [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+                    log.data($('AppSettings              :'), data.resource.properties.appSettings);
+                    log.data($('Use 32bit Worker Process :'), data.resource.properties.use32BitWorkerProcess);
+                    log.data('');
+                }
+            });
+        });
+
+
+
+    config.command('update [resource-group] [name]')
```
Comment (amarzavery) [inline lib/commands/arm/webapp/webapp.config._js]: command name should be `set`

File: lib/commands/arm/webapp/webapp.config._js [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+        .description($('Update webapp configuration'))
+        .usage('[options] <resource-group> <name>')
+        .option('-g --resource-group <resource-group>', $('the name of the resource group'))
+        .option('-n --name <name>', $('the name of the webapp to update'))
+        .option('-l --location <location>', $('the location of the webapp to update'))
+        .option('-j --jsonInput <jsonInput>', $('the json config object string'))
```
Comment (amarzavery) [inline lib/commands/arm/webapp/webapp.config._js]: should the json config object and other options be mutually exclusive?

File: lib/commands/arm/webapp/webapp.config._js [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+        .option('--phpversion <phpversion>', $('phpversion of webapp'))
+        .option('--pythonversion <pythonversion>', $('pythonversion of webapp'))
+        .option('--nodeversion <nodeversion>', $('nodeversion of webapp'))
+        .option('--appcommandline <appcommandline>', $('app command line for webapp'))
+        .option('--javaversion <javaversion>', $('javaversion of webapp'))
+        .option('--javacontainer <javacontainer>', $('javacontainer of webapp'))
+        .option('--javacontainerversion <javacontainerversion>', $('javacontainerversion of webapp'))
+        .option('--appsettings <appsettings>', $('webapp appsettings'))
```
Comment (amarzavery) [inline lib/commands/arm/webapp/webapp.config._js]: can multiple app settings be provided? If yes, then how? Please provide an accurate example in the help

File: lib/commands/arm/webapp/webapp.config._js [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+                jsonProperties['appCommandLine'] = options.appcommandline;
+            }
+
+            if (options.location !== null && options.location !== undefined) {
+                jsonObj['location'] = options.location;
+            }
+
+            if (options.appsettings !== null && options.appsettings !== undefined) {
+                var appSettingObj = {};
+                console.log('appsettings = ' + options.appsettings);
```
Comment (amarzavery) [inline lib/commands/arm/webapp/webapp.config._js]: please remove console.log statements from every where

File: lib/commands/arm/webapp/webapp.config._js [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+                var appSettingsSplit = options.appsettings.split(",");
+                for (var i = 0; i < appSettingsSplit.length; i++) {
+                    console.log(' appsettingStr = ' + appSettingsSplit[i]);
```
Comment (amarzavery) [inline lib/commands/arm/webapp/webapp.config._js]: remove this

File: lib/commands/arm/webapp/webapp.config._js [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+            var result;
+            try {
+                result = client.webSites.updateConfiguration(resourceGroup, name, options.slot, jsonObj, _);
+            } finally {
+                progress.end();
+            }
+            log.info('Webapp ' + webSiteSlotName + ' configuration has been updated ');
+        });
+
+	publish.command('get [resource-group] [name]')
```
Comment (amarzavery) [inline lib/commands/arm/webapp/webapp.config._js]: indentation is incorrect. The command name should be `show`

File: test/commands/arm/appserviceplan/arm.appserviceplan.feature-tests.js [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+var updatedPHPValue = "7.0";
+var servicePlanSku = "Standard";
+var numberOfWorkers = 1;
+var workerSize = 2;
+
+describe('arm', function () {
+    var suite;
+
+    before(function (done) {
+        suite = new CLITest(this, testPrefix);
```
Comment (amarzavery) [inline test/commands/arm/appserviceplan/arm.appserviceplan.feature-tests.js]: please set the tab settings to 2 spaces and replaces tabs with spaces for the following languages javascript, node.js and json in VS.

File: test/commands/arm/appserviceplan/arm.appserviceplan.feature-tests.js [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+        it('delete should work', function (done) {
+            suite.execute('appserviceplan delete %s %s --quiet --json', groupName, appservicename, function (result) {
+                result.exitStatus.should.equal(0);
+                done();
+            });
+        });
+
+    });
+
+    function createGroupAndPlan(done) {
```
Comment (amarzavery) [inline test/commands/arm/appserviceplan/arm.appserviceplan.feature-tests.js]: why do you need these methods? You have commands to create the appservice plans. Please remove all the helper methods.

File: test/commands/arm/appserviceplan/arm.appserviceplan.feature-tests.js [hunk 18]
```diff
@@ -0,0 +1,156 @@
...
+describe('arm', function () {
+    var suite;
+
+    before(function (done) {
+        suite = new CLITest(this, testPrefix);
+        suite.setupSuite(function () {
+            appservicename = suite.generateId('appservicepplanclitest', createdSites);
+            groupName = suite.generateId('testrg1', createdGroups);
+            var subscription = profile.current.getSubscription();
+            resourceClient = utils.createResourceClient(subscription);
```
Comment (amarzavery) [inline test/commands/arm/appserviceplan/arm.appserviceplan.feature-tests.js]: you dont need the above two lines

**Comment by amarzavery** [comment]
Please use the 0.11.0 version of azure-arm-website npm package. The current package that you are using is the old version and has a completely different style of autogenerated source code in it.

File: lib/commands/arm/appserviceplan/appserviceplan._js (hunk 1 above)
Comment (LukaszStem) [inline lib/commands/arm/appserviceplan/appserviceplan._js]: For aggregating those values into an object that will then be passed into this function call.
"about 15 lines lower"
`result = client.webHostingPlans.createOrUpdate(resourceGroup, parameters, _);`

File: lib/commands/arm/webapp/webapp.config._js (hunk 12 above)
Comment (LukaszStem) [inline lib/commands/arm/webapp/webapp.config._js]: Only one setting at a time for now. I've updated the description.

File: lib/commands/arm/appserviceplan/appserviceplan._js [hunk 19]
```diff
@@ -0,0 +1,251 @@
...
+    .option('-n --name <name>', $('the name of the App Service Plan to create'))
+    .option('-l --location <location>', $('the geographic region to create the App Service Plan, \n\t\t\t\t\t   locations found at https://azure.microsoft.com/en-us/status/'))
+    .option('-t --tier <tier>', $('tier of the App Service Plan' + tierInfo.toString()))
+    .option('-i --instances <numberOfInstances>', $('the number of instances in the App Service Plan, Default is 1'))
+    .option('-s, --subscription <id>', $('the subscription identifier'))
+    .execute(function(resourceGroup, name, location, tier, options, _) {
+      var subscription = profile.current.getSubscription(options.subscription);
+      var client = appserviceplanUtils.createWebappManagementClient(subscription);
+
+	    var numOfInstances;
```
Comment (haocs) [inline lib/commands/arm/appserviceplan/appserviceplan._js]: Please correct indentation.

File: lib/commands/arm/appserviceplan/appserviceplan._js [hunk 20]
```diff
@@ -0,0 +1,251 @@
...
+        return cli.missingArgument('name');
+      }
+      if (!location) {
+        return cli.missingArgument('location');
+      }
+      if (!tier) {
+        return cli.missingArgument('tier');
+      }
+      
+	    if('instances' in options) {
```
Comment (haocs) [inline lib/commands/arm/appserviceplan/appserviceplan._js]: Please correct indentation.

File: lib/commands/arm/appserviceplan/appserviceplan._js [hunk 21]
```diff
@@ -0,0 +1,251 @@
...
+	      numOfInstances = Number(options.instances);
+	    }
+	    else
+	    {
+	      numOfInstances = 1; //Default is 1
+	    }
+      var progress = cli.interaction.progress(util.format($('Creating appserviceplan %s'), name));
+      var tierUpper = tier.toUpperCase();
+      switch(tierUpper) {
+        case 'FREE':
```
Comment (haocs) [inline lib/commands/arm/appserviceplan/appserviceplan._js]: Please correct indentation.

File: lib/commands/arm/appserviceplan/appserviceplan._js [hunk 22]
```diff
@@ -0,0 +1,251 @@
...
+      case 'P1':
+      case 'P2':
+      case 'P3':
+        sku = 'Premium';
+        workerSize = tier.charAt(1) - 1;
+        break;
+      default:
+        log.info($('Invalid Tier'));
+      }
+      var envelope = {
```
Comment (haocs) [inline lib/commands/arm/appserviceplan/appserviceplan._js]: Please correct indentation.

File: lib/commands/arm/appserviceplan/appserviceplan._js [hunk 23]
```diff
@@ -0,0 +1,251 @@
...
+        progress.end();
+      }
+      log.info(util.format($('appserviceplan %s has been deleted'), name));
+    });
+
+  appserviceplan.command('list [resource-group]')
+    .description($('List your App Service Plans'))
+    .usage('[options] <resource-group>')
+    .option('-g --resource-group <resource-group>', $('the name of the resource group'))
+	  .option('-s --subscription <subscription>', $('the subscription identifier'))
```
Comment (haocs) [inline lib/commands/arm/appserviceplan/appserviceplan._js]: Please correct indentation.

File: lib/commands/arm/appserviceplan/appserviceplan._js [hunk 24]
```diff
@@ -0,0 +1,251 @@
...
+          log.data($('Number of workers  :'), data.sku.capacity);
+          log.data($('Worker Size        :'), translateWorkerSize(data.sku.name));
+          log.data('');
+        }
+      });
+    });
+
+	function translateWorkerSize(skuName)
+	{
+		var workerSize = skuName.charAt(1);
```
Comment (haocs) [inline lib/commands/arm/appserviceplan/appserviceplan._js]: Please correct indentation.

File: lib/commands/arm/webapp/webapp.config._js [hunk 25]
```diff
@@ -0,0 +1,194 @@
...
+'use strict';
+
+var profile = require('../../../util/profile');
+var utils = require('../../../util/utils');
+var webappUtils = require('./webappUtils');
+
+var $ = utils.getLocaleString;
+
+exports.init = function (cli) {
+    var log = cli.output;
```
Comment (haocs) [inline lib/commands/arm/webapp/webapp.config._js]: Please correct indentation. We use 2 spaces for indentation.

File: lib/commands/arm/webapp/webapp.config._js [hunk 26]
```diff
@@ -0,0 +1,194 @@
...
+                    log.data($('Java Container Version   :'), data.javaContainerVersion);
+                    log.data($('AlwaysOn                 :'), data.alwaysOn);
+                    log.data($('WebSockets Enabled       :'), data.webSocketsEnabled);
+                    log.data($('AppSettings              :'), data.appSettings);
+                    log.data($('Use 32bit Worker Process :'), data.use32BitWorkerProcess);
+                    log.data('');
+                }
+            });
+        });
+    config.command('update [resource-group] [name]')
```
Comment (haocs) [inline lib/commands/arm/webapp/webapp.config._js]: Please use `set` instead `update` to follow the naming convention.

File: lib/commands/arm/webapp/webapp.config._js [hunk 27]
```diff
@@ -0,0 +1,194 @@
...
+            if (options.javacontainer !== null && options.javacontainer !== undefined) {
+                jsonObj['javaContainer'] = options.javacontainer;
+            }
+            if (options.javacontainerversion !== null && options.javacontainerversion !== undefined) {
+                jsonObj['javaContainerVersion'] = options.javacontainerversion;
+            }
+            if (options.appcommandline !== null && options.appcommandline !== undefined) {
+                jsonObj['appCommandLine'] = options.appcommandline;
+            }
+			      var result;
```
Comment (haocs) [inline lib/commands/arm/webapp/webapp.config._js]: Please correct indentation.

File: lib/util/utils.js [hunk 28]
```diff
@@ -239,8 +239,8 @@ exports.createWebSiteExtensionsClient = function (siteName, hostNameSuffix, user
 };
 
 exports.createWebappManagementClient = function (subscription) {
-  var factoryMethod = require('azure-arm-website').createWebSiteManagementClient;
-  return _createArmClient(factoryMethod, subscription);
+    var factoryMethod = require('azure-arm-website');
```
Comment (haocs) [inline lib/util/utils.js]: Please correct indentation.

File: test/testlist-arm.txt [hunk 29]
```diff
@@ -1,98 +1,101 @@
...
-commands/arm/iothub/arm.iothub-tests.js
-#please put the following group tests to the last, as it uses the different subscription
-#and it exposes a test framework bug which failed vm tests randomly
-commands/arm/group/arm.group-tests.js
-commands/arm/group/arm.group.deployment-tests.js
-commands/arm/group/arm.group.deployment.operation-tests.js
-commands/arm/cdnMangement/arm.cdnMangement-tests.js
-commands/arm/policy/arm.policy.definition-tests.js
-commands/arm/policy/arm.policy.assignment-tests.js
+#commands/arm/powerbiembedded/arm.powerbiembedded-tests.js
```
Comment (haocs) [inline test/testlist-arm.txt]: Forgot to uncomment those tests? :)

**Comment by haocs** [comment]
@LukaszStem  I leave some comments. Most of them are about style issues. please use 2 spaces for indentation and follow the CLI naming convention. 
Also, please address the merge conflicts.

File: test/testlist-arm.txt (hunk 29 above)
Comment (LukaszStem) [inline test/testlist-arm.txt]: Yes..

File: lib/commands/arm/appserviceplan/appserviceplan._js (hunk 22 above)
Comment (LukaszStem) [inline lib/commands/arm/appserviceplan/appserviceplan._js]: This is resolved.

**Pull request merged**